### DISCUSSION TITLE
Didn't able to achieve parallelism using async and await in fastapi

### DISCUSSION THREAD

**Discussion by dummyuser-123** [fastapi/fastapi#11110] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, File, UploadFile, Form, Response
import uvicorn
import io
import numpy
import ultralytics
from ultralytics import YOLO
import cv2
import numpy as np
from PIL import Image
import asyncio

app = FastAPI()

async def detection(file):

	contents = file.file.read()

	temp_file_path = f"temp_image/{file.filename}"
	with open(temp_file_path, "wb") as temp_file:
	    temp_file.write(contents)

	model = YOLO('D:/GAN/yolo/models/yolov8n.pt')
	results = model.predict(source=temp_file_path, conf=0.25)

	for result in results:
	    boxes = result.boxes

	bbox=boxes.xyxy.tolist()[0]
	bbox = list(map(int, bbox))

	img = cv2.imread(temp_file_path)
	rgb_image = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)

	x, y, w, h = bbox
	color = (0, 255, 0)
	thickness = 2
	w = w - 100
	h = h - 100
	cv2.rectangle(rgb_image, (x, y), (x + w, y + h), color, thickness)

	pil_image = Image.fromarray(rgb_image)
	img_byte_io = io.BytesIO()
	pil_image.save(img_byte_io, format="PNG")
	img_byte_io.seek(0)
	output = img_byte_io.getvalue()

	return output

@app.post("/yolo/")
async def object_detection(file: UploadFile = File(...)):

	result = await detection(file)
	return Response(content=result, media_type="image/png")
```

### Description

I have created a simple object detection API using YOLO to test multiple parallel request handling asynchronously. I aim to examine whether, when three parallel requests are made simultaneously, all the requests receive responses concurrently (with no specific time constraint, but parallelism is crucial). After reading the official documentation of FastAPI, I attempted to implement async and await to support asynchronous operations in the API. However, when I try to make three parallel requests from different PCs using Insomnia (an alternative to Postman), the requests are processed sequentially rather than in parallel. For example, if request1 receives a response in 2 seconds, request2 in 4 seconds, and request3 in 6 seconds. 

I have also experimented with multithreading and multi-workers (using the uvicorn command), but unfortunately, I have not achieved success with these approaches either. Is there a mistake in the code, or is there another concept I may be missing in the API? I lack expertise in creating APIs, so if anyone knows how to achieve parallelism using async and await, please provide assistance.

### Operating System

Windows

### FastAPI Version

0.109.0

### Pydantic Version

2.6.0

### Python Version

3.10.6

**Comment by n8sty** [comment]
Try running your code without using `async` and `await` and you should get parallelism. Why? None of the methods within your functions are `async` so when a request is made that runs `detection` every line within that method blocks the event loop until the full function is finished.

**Comment by jgould22** [comment, marked as answer]
To expand on the sibling comment.

Python async is [co-operative multitasking](https://en.wikipedia.org/wiki/Cooperative_multitasking) which means that only one task on the event loop is running a time.

You want to keep the amount of code between any 2 await functions to a minimum.  The model and image work you are doing is not suitable to an async environment as it is more CPU bound work. 

In the short term you can change your `object_detection` function to a def and FastAPI will spawn it on a thread pool for you and not block the event loop.

**Comment by dummyuser-123** [comment]
Thank you @n8sty and @jgould22 for the solution, finally after a lot of struggle I have achieved the parallelism.

**Comment by dummyuser-123** [comment]
@jgould22 I am able to achieve parallelism for all requests but I am facing some strange behavior. Like if one requests is made then it is taking 3 seconds to give output and if 3 parallel requests are made then it is taking 9 seconds for each user to get the response. So here the time increases as the number of requests increases. I have tried ProcessPoolExecutor and run_in_threadpool() but both are giving me same output. So, do know how to solve this problem ?

**Reply by YuriiMotov** [reply]
Run your app under the `gunicorn` with multiple workers: https://fastapi.tiangolo.com/deployment/server-workers/

**Reply by dummyuser-123** [reply]
But I have initially loaded the model in the memory for faster inference. So, if I increase the number of workers then it will take huge amount of memory to handle the requests. Because here each worker will have an separate copy of the model to process the request. And this is cost ineffective way. I am finding such solution where from single worker only I can able to handle some amount of requests.

**Reply by YuriiMotov** [reply]
You should understand that all your endpoint calls are executed in separate threads, but inside one process. So, they share CPU time of this process. The more threads you run, the more often CPU will switch from one thread to another.
To increase speed you need to run your CPU-bound tasks in separate processes, so they can use different CPU-cores and this will increase speed.
If you can't load model for every worker, you should somehow share this model between processes.
Take a look at this module: https://docs.python.org/3/library/multiprocessing.shared_memory.html
I'm not sure, but it seems it can help you.

**Reply by dummyuser-123** [reply]
I have used ProcessPoolExecutor from concurrent.futures to run my cpu-bound model into separate process but then my api endpoint is scheduling the requests in sequentially manner. Means it is blocking the event loop. So, how to faced such problem with multiprocessing ?

**Reply by YuriiMotov** [reply]
The easiest (but not very efficient) way is to make endpoint functions not async.
More efficient way requires efforts to figure out how to run processes in async functions and await for their results. And combine this with sharing memory between processes.
That's really interesting question. I'll try to find time to figure out and create an example next week.

I think you can also ask this question (how to run processes from async code and share memory between these processes) on stackowerflow, since it's not a FastAPI-related question

**Reply by dummyuser-123** [reply]
Sure I will ask this question on stackoverflow. By that time, if you find any solution regarding this problem then please share the solution. And btw thanks for the help.

**Reply by YuriiMotov** [reply]
@dummyuser-123 , I found some time to figure this out. I'm not sure it's a good solution, but it's all I managed to create so far.
Using this approach you have to ensure that `Model.process` method is thread-safe (doesn't modify Model object data or uses special mechanisms to synchronize writing access).
```
import asyncio
import random
import time
from contextlib import asynccontextmanager
from multiprocessing import Pool, cpu_count
from multiprocessing.pool import Pool as MultiprocPool
from typing import Annotated, cast

from fastapi import Depends, FastAPI, Request


class Model:
    def __init__(self):
        print("Model init started")
        self._data = []
        for _ in range(5_000_000):
            self._data.append(random.randint(1, 10000))
        print("Model init finished")

    def process(self, prompt: str):
        print(f"process({prompt=}) started at {time.time()}")
        res = 0
        for _ in range(5):
            for n in self._data:
                res += n % 2
        print(f"process({prompt=}) finished at {time.time()}")
        return {"prompt": prompt, "res": res}


@asynccontextmanager
async def lifespan(app: FastAPI):
    model = Model()     # Load model
    pool_size = cpu_count() * 2
    with Pool(processes=pool_size) as pool:
        yield {"model": model, "proc_pool": pool}


app = FastAPI(lifespan=lifespan)


def get_model(request: Request) -> Model:
    return cast(Model, request.state.model)


def get_proc_pool(request: Request) -> MultiprocPool:
    return cast(MultiprocPool, request.state.proc_pool)


@app.get("/")
async def index(
    model: Annotated[Model, Depends(get_model)],
    proc_pool: Annotated[MultiprocPool, Depends(get_proc_pool)],
    prompt: str,
):
    t_start = time.time()
    print(f"Processing request ({prompt=})")
    res = proc_pool.apply_async(model.process, (prompt,))
    timeout = 1000
    while timeout > 0:
        if res.ready():
            break
        await asyncio.sleep(0.1)
        timeout -= 1

    if res.ready:
        print(
            f"Request has processed ({prompt=}). Execution time: {time.time() - t_start}"
        )
        return res.get()
    else:
        print(
            f"Request processing failed ({prompt=}). Execution time: {time.time() - t_start}"
        )

    return {}
```

**Reply by YuriiMotov** [reply]
@dummyuser-123, have you managed to solve your problem?
